### PR TITLE
Constraint on the form of term

### PR DESCRIPTION
**Pull request by Kakadu**
Constraint on the form of term. Invented during discussion about pattern matching compilation.

### CODE REVIEW

File: src/core/Core.ml [hunk 1]
```diff
@@ -205,6 +226,12 @@ let diseq = (=/=)
 let delay g st = RStream.from_fun (fun () -> g () st)
 
 let conj f g st = RStream.bind (f st) g
+
+let structural var rr k st =
+  match Term.var var with
+  | None -> success st
+  | Some v -> success { st with State.prunes = Term.VarMap.add v (Obj.magic (rr,k)) (State.prunes st) }
```
Comment (eupp) [inline src/core/Core.ml:233]: Shouldn't you check the constraint when you first add it? 
What if it is violated in the current substitution?

File: src/core/Core.mli [hunk 2]
```diff
@@ -42,6 +42,19 @@ val unify : ('a, 'b logic) injected -> ('a, 'b logic) injected -> goal
 (** [x =/= y] creates a goal, which introduces a disequality constraint for [x] and [y] *)
 val (=/=) : ('a, 'b logic) injected -> ('a, 'b logic) injected -> goal
 
+(* Call [structural var reifier checker] adds a structural constraint for future use.
+ * Every time substitution is updated it reifies [var] using [reifier] and checks that
+ * the result satisfies desired predicate [checker].
+ *
+ * The predicate [checker] returns true when constraint is violated.
```
Comment (eupp) [inline src/core/Core.mli:49]: I, personally, would prefer the opposite (return false when the constraint is violated).

File: src/core/Core.ml (hunk 1 above)
Comment (Kakadu) [reply src/core/Core.ml]: fixed.

**Comment by eupp** [comment]
The idea to have boolean constraints on terms that prune the search space is great. 

However, I have a concern about the current implementation. If I understood correctly, these new constraints are not attached to some particular variable and they are completely rechecked after each update of the substitution. I'm afraid it can scale poorly for large problems. 
It would be cool if these constraints could be checked "incrementally" as the substitution refines, 
and if we can attach each constraint to some particular variable (or some set of variables) and recheck the constraint only if the related variables have changed. 

Is a performance concern to you, or the naive implementation works fine for your problems?

**Comment by Kakadu** [comment]
> However, I have a concern about the current implementation. If I understood correctly, these new constraints are not attached to some particular variable and they are completely rechecked after each update of the substitution. I'm afraid it can scale poorly for large problems.
> It would be cool if these constraints could be checked "incrementally" as the substitution refines,
> and if we can attach each constraint to some particular variable (or some set of variables) and recheck the constraint only if the related variables have changed.

I thought about that and I understand how it can be done for idempotent substitution. But for triangular one every extension of substitution can change all keys in the substitution so we kind of obliged to check all constraints. 

> Is a performance concern to you, or the naive implementation works fine for your problems?

Without these constraints it doesn't work at all :)

**Comment by eupp** [comment]
I have an idea of how it can be done. 
But it will require to write constraints in the CPS style. 
For example, the `even` constraint would become:

```OCaml
(* returns false when constraint is violated *)
let even xs =
  (* here `llist` is a logic list *)
  let rec helper ~iseven llist = 
    Logic.constraint llist (fun list -> 
       (* here we pull off one logic on the top, but `tl` is still a logic list *)
       match list with
       | Nil -> iseven
       | Cons (_, tl) -> helper ~iseven:(not iseven) tl
    )
  helper ~iseven:true xs
```

The idea is to have a continuation, that will be fired when the variable is assigned to some term constructor in the substitution. We attach the continuation to the variable.
Note that although the variable is assigned to the constructor, the the term still can have variables inside. In this case, we will run the continuations to inspect the term deeper, until we met a variable in the term. At this point, we again assign the new continuation to the new variable and wait for it to be instantiated. 

I hope you can grasp the idea from the explanation above. What do you think about it?

**Comment by Kakadu** [comment]
1st: it is too early to implement this optimization
2nd: you will have issues with reification of inner variable, because toplevel reifier will not work

**Comment by eupp** [comment]
> 1st: it is too early to implement this optimization
> 2nd: you will have issues with reification of inner variable, because toplevel reifier will not work

Ok then :)

**Pull request merged**